### PR TITLE
8309216: Cast from jchar* to char* in test java/io/GetXSpace.java

### PR DESCRIPTION
**Pull request by bplb**
For Unix, copy the `jchar`s into an allocated `char` array.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309216](https://bugs.openjdk.org/browse/JDK-8309216): Cast from jchar* to char* in test java/io/GetXSpace.java

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14276/head:pull/14276` \
`$ git checkout pull/14276`

Update a local copy of the PR: \
`$ git checkout pull/14276` \
`$ git pull https://git.openjdk.org/jdk.git pull/14276/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14276`

View PR using the GUI difftool: \
`$ git pr show -t 14276`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14276.diff">https://git.openjdk.org/jdk/pull/14276.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14276#issuecomment-1572943099)

### CODE REVIEW

**Comment by bplb** [comment]
Preliminary test run passes.

File: test/jdk/java/io/File/libGetXSpace.c [hunk 1]
```diff
@@ -124,9 +128,24 @@ Java_GetXSpace_getSpace0
...
+    if (chars == NULL) {
+        (*env)->ReleaseStringChars(env, root, strchars);
+        JNU_ThrowByNameWithLastError(env, "java/lang/RuntimeException",
+                                     "malloc");
+        return JNI_FALSE;
+    }
+
+    for (int i = 0; i < len; i++) {
+        chars[i] = (char)strchars[i];
+    }
```
Comment (RogerRiggs) [inline test/jdk/java/io/File/libGetXSpace.c:142]: Is this `memcpy(chars, strchars, len);`

File: test/jdk/java/io/File/libGetXSpace.c [hunk 2]
```diff
@@ -77,17 +77,19 @@ Java_GetXSpace_getSpace0
         }
     }
 
-    LPCWSTR path = (LPCWSTR)chars;
+    LPCWSTR path = (LPCWSTR)strchars;
 
     if (pfnGetDiskSpaceInformation != NULL) {
         // use GetDiskSpaceInformationW
         DISK_SPACE_INFORMATION diskSpaceInfo;
         BOOL hres = pfnGetDiskSpaceInformation(path, &diskSpaceInfo);
```
Comment (RogerRiggs) [inline test/jdk/java/io/File/libGetXSpace.c:85]: The strchars/path could be released before checking the return; keeping the cleanup in the common flow.
Here and below after calling GetDiskFreeSpaceExW.
Your call.

File: test/jdk/java/io/File/libGetXSpace.c (hunk 1 above)
Comment (bplb) [reply test/jdk/java/io/File/libGetXSpace.c:142]: I don't think so if `chars` is `char *` and `strchars` is `jchar*` as I think `char` is 8 bits and `jchar` is 16 bits.

File: test/jdk/java/io/File/libGetXSpace.c (hunk 2 above)
Comment (bplb) [reply test/jdk/java/io/File/libGetXSpace.c:85]: Good idea; will update.

Comment (bplb) [reply test/jdk/java/io/File/libGetXSpace.c:85]: So changed in 0ecf89b1d060bb8c1a35402ec108cbe9f3f35bb9.

File: test/jdk/java/io/File/libGetXSpace.c (hunk 1 above)
Comment (naotoj) [reply test/jdk/java/io/File/libGetXSpace.c:141]: I just wonder if jchar is UTF-16 and char is platform encoding, would it work by simply casting the low 8bit to char?

Comment (bplb) [reply test/jdk/java/io/File/libGetXSpace.c:141]: Probably not. This is only intended to cover the constrained case of this test. Something like the macro `WITH_PLATFORM_STRING` in `libjava/io_util.h` would be needed for the general case, but I thought that that might be overdoing it for here. I could of course be mistaken.

Comment (naotoj) [reply test/jdk/java/io/File/libGetXSpace.c:141]: OK, if the usage is limited and covers the required test cases, that sounds reasonable.

**Review by naotoj — APPROVED** [review]

File: test/jdk/java/io/File/libGetXSpace.c (hunk 1 above)
Comment (bplb) [reply test/jdk/java/io/File/libGetXSpace.c:141]: Only file system roots are passed as the paths handed to the `getSpace0()` method so hopefully it should be all right.

**Pull request closed without merging**